src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -94,18 +52,534 @@ impl fmt::Display for Mode {
+                    return true;
+                }
+            }
+
+            Rvalue::NullaryOp(NullOp::Box, _) => return true,
+
+            _ => {}
+        }
+
+        Self::in_rvalue_structurally(cx, rvalue)
+    }
+
+    fn in_call(
+        cx: &ConstCx<'_, 'tcx>,
+        callee: &Operand<'tcx>,
+        args: &[Operand<'tcx>],
+        _return_ty: Ty<'tcx>,
+    ) -> bool {
+        let fn_ty = callee.ty(cx.mir, cx.tcx);
+        match fn_ty.sty {
+            ty::FnDef(def_id, _) => {
+                match cx.tcx.fn_sig(def_id).abi() {
+                    Abi::RustIntrinsic |
+                    Abi::PlatformIntrinsic => {
+                        assert!(!cx.tcx.is_const_fn(def_id));
+                        match &cx.tcx.item_name(def_id).as_str()[..] {
+                            | "size_of"
+                            | "min_align_of"
+                            | "needs_drop"
+                            | "type_id"
+                            | "bswap"
+                            | "bitreverse"
+                            | "ctpop"
+                            | "cttz"
+                            | "cttz_nonzero"
+                            | "ctlz"
+                            | "ctlz_nonzero"
+                            | "overflowing_add"
+                            | "overflowing_sub"
+                            | "overflowing_mul"
+                            | "unchecked_shl"
+                            | "unchecked_shr"
+                            | "rotate_left"
+                            | "rotate_right"
+                            | "add_with_overflow"
+                            | "sub_with_overflow"
+                            | "mul_with_overflow"
+                            | "saturating_add"
+                            | "saturating_sub"
+                            | "transmute"
+                            => return true,
+
+                            _ => {}
+                        }
+                    }
+                    _ => {
+                        let is_const_fn =
+                            cx.tcx.is_const_fn(def_id) ||
+                            cx.tcx.is_unstable_const_fn(def_id).is_some() ||
+                            cx.is_const_panic_fn(def_id);
+                        if !is_const_fn {
+                            return true;
+                        }
+                    }
+                }
+            }
+            _ => return true,
+        }
+
+        Self::in_operand(cx, callee) || args.iter().any(|arg| Self::in_operand(cx, arg))
+    }
+}
+
+// Refers to temporaries which cannot be promoted as
+// promote_consts decided they weren't simple enough.
+struct IsNotPromotable;
+
+impl Qualif for IsNotPromotable {
+    const IDX: usize = 3;
+
+    fn in_call(
+        cx: &ConstCx<'_, 'tcx>,
+        callee: &Operand<'tcx>,
+        _args: &[Operand<'tcx>],
+        _return_ty: Ty<'tcx>,
+    ) -> bool {
+        if cx.mode == Mode::Fn {
```

Comment:
I was discussing `IsNotPromotable` with @oli-obk... it's not exactly clear what to make of it, but the names are definitely wrong.

Promotability is tricky because it differs between compile-time and run-time.

Ideally we would just get rid of `IsNotPromotable`, but at least one run-time thing `(`#[rustc_const_arg]`) is stable *and* differs in desired behavior from "promotion involving `const fn` calls".

Reply:
> it differs between compile-time and run-time.

I am not sure what you mean by this. Do you mean the analysis is different in const context?

Reply:
I mean we allow promoting `&const_fn()` in a `const`, given that interior mutability and drop conditions are met, while at runtime `const_fn` *also* needs `#[rustc_promotable]`.

Reply:
Okay so by "at runtime" you mean "in run-time code".

However, currently there are also a bunch of checks in `NotConst` that have to do with promotability, from what I can see.

Reply:
Like I said, some of the naming is wrong/misleading and should be changed.

IIRC, `IsNotConst` should have the `IsNotPromotable` name, and our current `IsNotPromotable` should be *...checks IRC log...* `IsNotImplicitlyPromotable` / `MayCallNotImplicitlyPromotableConstFn` / `MayCallUnpredictableConstFn`.

(I'm not sure what @oli-obk actually prefers)

Reply:
To expand a bit: the main point was that we should distinguish between:
* "implicit promotion"
  * only "rvalue-to-`'static`" AFAIK
  * requires `#[rustc_promotable]` on called `const fn`s
* "explicit promotion" / "forced promotion"
  * e.g. "`const` args" (whether `#[rustc_const_arg]` or a different system)
  * it can't compile without being promoted, so we can allow pretty much everything we do in a `const` item, modulo the "SSA tree" shape required to actually perform the promotion on MIR (we'd need something like VSDG to allow more)

We should figure out how to avoid this discussion getting lost though, heh.

Reply:
Maybe `HasNoConstFnCalls`? (I know it's wrong in the presence of `rustc_promotable`)